### ISSUE TITLE
rvierdiiev - Players are not guaranteed to receive rewards

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-01-derby-judging#143]
rvierdiiev

high

# Players are not guaranteed to receive rewards

## Summary
Players are not guaranteed to receive rewards. This is because, when player's rewards are accrued, they are not reserved, till he calls `redeemRewards`.
## Vulnerability Detail
When user calls `rebalanceBasket`, then rewards [are added](https://github.com/sherlock-audit/2023-01-derby/blob/main/derby-yield-optimiser/contracts/Game.sol#L396-L398) to his `totalUnRedeemedRewards` variable.
And this rewards are not reserved anyhow in the vault. They will be reserved, only after user calls `redeemRewards`, then game will message Vault [to reserve unredeemed rewards for that user](https://github.com/sherlock-audit/2023-01-derby/blob/main/derby-yield-optimiser/contracts/Game.sol#L552).

But in case if players, do not call `redeemRewards`, there is a risk that they will not receive their rewards.
For example, for some reasons all people withdraw from vaults on all chains and some players still didn't call `redeemRewards`, they will loose them as there will be no funds to cover their rewards as they are already withdrawn by depositors.

Also the problem can occur for last withdrawers in case of many players will call `redeemRewards` at same time. This will decrease vault share price with a big value, so last depositors will lose some funds and receive less funds for same shares as depositors who withdrawn before players called `redeemRewards` together.

Duplicate of #269

I think that current approach is not correct.
Suppose that vault has 1.1 million of underlying and 1 million shares for simplicity. So we see that 0.1 million were yields that vault accrued. So exchange rate is 1.1 now for anyone who wants to withdraw. But actually part of this 0.1 million belongs to the players that didn't withdraw yet. Suppose that this part is 0.05 million. That means that real exhange rate should be 1.05, not 1.1. As result because of this rate funds are distributed incorrectly.

I don't know which severity this should be. I started with thought about loosing of funds for last rewards redeemers, but ended with the point that exhange rate is incorrectly calculated, which constantly affects deposits/withdraws. I will set it as high and let judges decide.
## Impact
Players are not guaranteed to receive rewards and last withdrawers can receive less funds. Exchange rate is calculated incorrectly.
## Code Snippet
Provided above
## Tool used

Manual Review

## Recommendation
You need to think about the way, how to reserve rewards, once they accumulated. Maybe good solution will be to mint Vault lp tokens for rewards to players. Or you need to make vault reserve rewards, once they accrued and also think about that they can be negative, so in that case you should also be able to release reserved funds.
